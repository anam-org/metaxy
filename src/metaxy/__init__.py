from metaxy.models.container import ContainerDep, ContainerSpec, SpecialContainerDep
from metaxy.models.feature import Feature
from metaxy.models.feature_spec import FeatureDep, FeatureDepMetadata, FeatureSpec
from metaxy.models.types import ContainerKey, FeatureKey

__all__ = [
    "Feature",
    "FeatureDep",
    "FeatureDepMetadata",
    "FeatureSpec",
    "ContainerDep",
    "ContainerSpec",
    "SpecialContainerDep",
    "FeatureKey",
    "ContainerKey",
]
