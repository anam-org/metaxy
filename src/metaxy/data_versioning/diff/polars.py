"""Polars implementation of metadata diff resolver."""

import polars as pl

from metaxy.data_versioning.diff.base import DiffResult, MetadataDiffResolver


class PolarsDiffResolver(MetadataDiffResolver):
    """Identifies changed rows using Polars operations.

    Strategy:
    - Categorizes changes into added, changed, and removed
    - Uses LEFT/RIGHT JOINs to identify each category
    - Materializes once and splits into three DataFrames (efficient)
    """

    def find_changes(
        self,
        target_versions: pl.LazyFrame,
        current_metadata: pl.LazyFrame | None,
    ) -> DiffResult:
        """Find all changes between target and current.

        Args:
            target_versions: LazyFrame with calculated data_versions
            current_metadata: LazyFrame with current metadata, or None

        Returns:
            DiffResult with three materialized DataFrames
        """
        # Select only sample_id and data_version from target_versions
        # (it may have intermediate joined columns from upstream)
        target_versions = target_versions.select(["sample_id", "data_version"])

        if current_metadata is None:
            # No existing metadata - all target rows are new, nothing removed/changed
            empty_df = pl.DataFrame({"sample_id": [], "data_version": []})
            return DiffResult(
                added=target_versions.collect(),
                changed=empty_df,
                removed=empty_df,
            )

        # Join target with current to categorize changes
        # Keep only sample_id and data_version from current for comparison
        current_comparison = current_metadata.select(
            ["sample_id", pl.col("data_version").alias("__current_data_version")]
        )

        # LEFT JOIN target with current
        compared = target_versions.join(
            current_comparison,
            on="sample_id",
            how="left",
        )

        # Build lazy queries for each category
        added_lazy = (
            compared.filter(pl.col("__current_data_version").is_null())
            .drop("__current_data_version")
            .select(["sample_id", "data_version"])
        )

        changed_lazy = (
            compared.filter(
                pl.col("__current_data_version").is_not_null()
                & (pl.col("data_version") != pl.col("__current_data_version"))
            )
            .drop("__current_data_version")
            .select(["sample_id", "data_version"])
        )

        removed_lazy = current_metadata.join(
            target_versions.select("sample_id"),
            on="sample_id",
            how="anti",
        ).select(["sample_id", "data_version"])

        # Collect all three at once (single execution for shared computations)
        added_df, changed_df, removed_df = pl.collect_all(
            [added_lazy, changed_lazy, removed_lazy]
        )

        return DiffResult(
            added=added_df,
            changed=changed_df,
            removed=removed_df,
        )
