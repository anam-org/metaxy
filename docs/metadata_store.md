# MetadataStore Tutorial

This guide introduces Metaxy's `MetadataStore` component, explaining its design philosophy, capabilities, and common usage patterns.

## What is a MetadataStore?

A `MetadataStore` is responsible for persisting and retrieving feature metadata - the information about your samples, their data versions, and any additional attributes you want to track.

Think of it as a versioned database for your feature pipeline, where each row represents a sample, and the metadata tracks which version of processing produced that sample.

## Core Concepts

### Metadata DataFrame

Metadata is represented as a Polars DataFrame with a specific structure:

```python
import polars as pl

metadata_df = pl.DataFrame({
    "sample_id": [1, 2, 3],
    "path": ["/data/sample1.mp4", "/data/sample2.mp4", "/data/sample3.mp4"],
    "data_version": [
        {"frames": "abc123", "audio": "def456"},
        {"frames": "abc124", "audio": "def457"},
        {"frames": "abc125", "audio": "def458"},
    ]
})
```

Key characteristics:
- **Required column**: `data_version` (pl.Struct type) - Maps container names to their version hashes
- **Sample identifiers**: Typically `sample_id` or similar columns to uniquely identify samples
- **User-defined columns**: Any additional metadata (paths, labels, timestamps, etc.)

### Immutability

**All metadata is immutable.** Once written, it never changes.

Why?
- **Reproducibility**: You can always recreate the exact same data
- **Caching**: Safe to cache aggressively since data never changes
- **Concurrency**: Multiple processes can write simultaneously without conflicts
- **Auditing**: Complete history is preserved

This means:
- ✅ Writing new metadata is always append-only
- ✅ Same data_version = same data, guaranteed
- ❌ No update or overwrite operations
- ❌ No delete operations (by design)

### Features and Containers

Metadata is organized by **features** (top-level entities) and **containers** (sub-components of features).

Example: A video processing feature might have:
```python
class VideoFeature(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "processing"]),
    containers=[
        ContainerSpec(key=ContainerKey(["frames"]), code_version=1),
        ContainerSpec(key=ContainerKey(["audio"]), code_version=2),
    ]
)):
    pass
```

The `data_version` struct has one field per container:
```python
{"frames": "hash_of_frames_v1", "audio": "hash_of_audio_v2"}
```

## Design Philosophy

### Why Immutable?

In data pipelines, **reproducibility is paramount**. When you see a specific `data_version`, you should be able to:
1. Recreate the exact same output
2. Debug issues by looking at historical versions
3. Roll back to previous versions if needed
4. Share and compare results across environments

Immutability makes all of this trivial - there's no "current state" that might change, only a growing collection of versioned snapshots.

### Why Composable Stores?

Real-world workflows often span multiple environments:
- **Development**: Testing new features locally
- **Staging**: Validating changes before production
- **Production**: Running stable, battle-tested code

With composable stores, you can:
- Write to your local/development store
- Read unchanged features from production (via fallback)
- Only recompute what's actually changed
- Keep environments isolated (writes never affect upstream stores)

This is especially powerful for **branch deployments** where you're testing changes to one feature while reusing upstream features from production.

## Key Capabilities

### 1. Write Metadata

```python
from metaxy.metadata_store import InMemoryMetadataStore
import polars as pl

store = InMemoryMetadataStore()

# Prepare metadata with data_version
metadata = pl.DataFrame({
    "sample_id": [1, 2, 3],
    "path": ["/data/1.mp4", "/data/2.mp4", "/data/3.mp4"],
    "data_version": [
        {"frames": "abc123", "audio": "def456"},
        {"frames": "ghi789", "audio": "jkl012"},
        {"frames": "mno345", "audio": "pqr678"},
    ]
})

# Write to store (immutable, append-only)
store.write_metadata(VideoFeature, metadata)
```

### 2. Read Metadata

```python
# Read all metadata for a feature
df = store.read_metadata(VideoFeature)

# Read with filters
df = store.read_metadata(
    VideoFeature,
    filters=pl.col("sample_id") > 1
)

# Read specific columns only
df = store.read_metadata(
    VideoFeature,
    columns=["sample_id", "data_version"]
)
```

### 3. Check Feature Existence

```python
# Check if feature exists locally
if store.has_feature(VideoFeature):
    df = store.read_metadata(VideoFeature)

# Check including fallback stores
if store.has_feature(VideoFeature, check_fallback=True):
    df = store.read_metadata(VideoFeature, allow_fallback=True)
```

### 4. Fallback Store Chain

Create a store that reads from multiple sources:

```python
# Production store (read-only from dev's perspective)
prod_store = InMemoryMetadataStore()

# Development store with prod as fallback
dev_store = InMemoryMetadataStore(
    fallback_stores=[prod_store]
)

# This tries dev_store first, then prod_store
df = dev_store.read_metadata(VideoFeature, allow_fallback=True)

# This only checks dev_store
df = dev_store.read_metadata(VideoFeature, allow_fallback=False)
```

### 5. Upstream Dependency Resolution

Automatically load all upstream dependencies for a feature:

```python
class ProcessedVideoFeature(Feature, spec=FeatureSpec(
    key=FeatureKey(["processed", "video"]),
    deps=[FeatureDep(key=FeatureKey(["video", "processing"]))],
    containers=[
        ContainerSpec(
            key=ContainerKey(["default"]),
            code_version=1,
            deps=[ContainerDep(
                feature_key=FeatureKey(["video", "processing"]),
                containers=[ContainerKey(["frames"]), ContainerKey(["audio"])]
            )]
        )
    ]
)):
    pass

# Get all upstream metadata as dict
upstream_metadata = store.read_upstream_metadata(ProcessedVideoFeature)
# Returns: {"video_processing": <DataFrame with data_version column>}
```

### 6. Data Version Calculation

Automatically calculate data versions using upstream dependencies:

```python
# New samples to process (without data_version yet)
new_samples = pl.DataFrame({
    "sample_id": [4, 5, 6],
    "path": ["/data/4.mp4", "/data/5.mp4", "/data/6.mp4"],
})

# Calculate data versions based on upstream metadata and write
result_df = store.calculate_and_write_data_versions(
    feature=ProcessedVideoFeature,
    sample_df=new_samples,
    allow_upstream_fallback=True  # Can read upstream from fallback stores
)

# result_df now has data_version column calculated via Merkle tree hashing
print(result_df["data_version"])
```

## Computation Strategies

The `MetadataStore` supports two computation strategies for calculating data versions:

### 1. Backend-Native Computation (Optimized)

If the storage backend supports it AND all upstream dependencies are available locally, the store can compute data versions using native operations (e.g., SQL with SHA256 functions).

**Requirements:**
- Backend implements `_compute_data_versions_native()` method
- All upstream features exist in the same store (checked via `has_feature()`)
- No fallback to other stores needed

**Benefits:**
- Maximum performance (no data movement)
- Leverages database optimizations (parallelism, indexes)
- Memory efficient (computation happens in-database)

Example (conceptual - DuckDB backend):
```sql
-- Pure SQL computation with native SHA256
SELECT 
    new_samples.*,
    struct_pack(
        default := sha256_hex(
            concat('default|1|upstream_feature/frames|', 
                   upstream.data_version.frames)
        )
    ) as data_version
FROM new_samples
LEFT JOIN upstream_feature USING (sample_id)
```

### 2. Polars Computation (Universal Fallback)

If native computation isn't possible (backend doesn't support it OR upstream data is in fallback stores), the store falls back to Polars-based computation.

**Process:**
1. Load all upstream metadata into memory as Polars DataFrames
2. Use `polars-hash` plugin for SHA256 hashing
3. Compute data versions using Polars expressions
4. Write results back to store

**Benefits:**
- Works everywhere (universal fallback)
- Consistent hashing across all backends
- Handles cross-store scenarios gracefully

The strategy selection is **automatic and transparent** - you don't need to think about it!

```python
# This automatically chooses the best strategy:
# - Native if possible (all deps local, backend supports it)
# - Polars otherwise (deps in fallback stores or backend doesn't support native)
result = store.calculate_and_write_data_versions(
    feature=MyFeature,
    sample_df=new_samples,
)
```

## Common Usage Patterns

### Pattern 1: Simple Single-Store Usage

For local development or testing:

```python
from metaxy.metadata_store import InMemoryMetadataStore

store = InMemoryMetadataStore()

# Write metadata
store.write_metadata(MyFeature, metadata_df)

# Read it back
df = store.read_metadata(MyFeature)

# Check if feature exists
if store.has_feature(MyFeature):
    df = store.read_metadata(MyFeature)
else:
    # Compute and materialize
    df = compute_feature(MyFeature)
    store.write_metadata(MyFeature, df)
```

### Pattern 2: Branch Deployment Workflow

Testing changes while reusing production data:

```python
# Setup: Production store (already has FeatureA materialized)
prod_store = InMemoryMetadataStore()
# ... prod_store has FeatureA metadata ...

# Setup: Dev store with prod fallback
dev_store = InMemoryMetadataStore(fallback_stores=[prod_store])

# Scenario: Testing FeatureB (depends on FeatureA)
# FeatureA hasn't changed, so we can reuse from prod
# FeatureB has new code version, needs recomputation

# This will:
# 1. Check if all upstream (FeatureA) is in dev_store - NO
# 2. Fall back to Polars computation (loads FeatureA from prod_store)
# 3. Calculate FeatureB data_versions using Polars
# 4. Write FeatureB to dev_store only (prod remains unchanged)

new_feature_b_samples = pl.DataFrame({
    "sample_id": [1, 2, 3],
    "custom_field": ["a", "b", "c"],
})

result = dev_store.calculate_and_write_data_versions(
    feature=FeatureB,
    sample_df=new_feature_b_samples,
    allow_upstream_fallback=True  # Key: enables reading from prod
)

# Now dev_store has FeatureB, prod_store is untouched
assert dev_store.has_feature(FeatureB, check_fallback=False)  # True
assert prod_store.has_feature(FeatureB, check_fallback=False)  # False
```

### Pattern 3: Multi-Environment Pipeline

Chaining multiple environments:

```python
# Three-tier setup
prod_store = InMemoryMetadataStore()
staging_store = InMemoryMetadataStore(fallback_stores=[prod_store])
dev_store = InMemoryMetadataStore(fallback_stores=[staging_store])

# Dev can read from staging → prod
# Staging can read from prod
# Each writes only to itself

# Promotion workflow:
# 1. Develop and test in dev_store
# 2. Promote to staging_store (explicit copy)
# 3. Validate in staging
# 4. Promote to prod_store

# Promotion is just reading from dev and writing to staging
dev_metadata = dev_store.read_metadata(MyFeature, allow_fallback=False)
staging_store.write_metadata(MyFeature, dev_metadata)
```

### Pattern 4: Incremental Processing

Only process new samples by filtering existing metadata:

```python
# Get all previously processed samples
existing = store.read_metadata(MyFeature, columns=["sample_id"])
existing_ids = set(existing["sample_id"].to_list())

# Filter to only new samples
all_samples = pl.DataFrame({
    "sample_id": [1, 2, 3, 4, 5],
    "path": ["...", "...", "...", "...", "..."],
})

new_samples = all_samples.filter(
    ~pl.col("sample_id").is_in(existing_ids)
)

if len(new_samples) > 0:
    # Process and write only new samples
    result = store.calculate_and_write_data_versions(
        feature=MyFeature,
        sample_df=new_samples
    )
```

## Storage Backends

### InMemoryMetadataStore

Simple dict-based storage, useful for:
- Testing and prototyping
- Single-process applications
- Small datasets that fit in memory

```python
from metaxy.metadata_store import InMemoryMetadataStore

store = InMemoryMetadataStore()
```

**Limitations:**
- Data lost when process exits
- Not suitable for production
- No persistence across runs
- Uses Polars computation only (no native optimization)

### Future Backends (see roadmap.md)

The abstract `MetadataStore` interface enables future implementations:
- **Delta Lake**: Production-grade with time travel, ACID transactions
- **DuckDB**: High-performance analytical queries with native computation
- **ClickHouse**: Distributed analytics with SQL-based computation
- **SQLAlchemy**: Traditional database storage (generic SQL support)

Each backend can implement **backend-specific optimizations** like:
- **Native computation**: SHA256 in SQL instead of Polars
- **Partitioning**: Physical data layout (e.g., by `data_version`)
- **Indexing**: Fast lookups on specific columns
- **Query pushdown**: Execute filters at storage layer

## Partitioning Strategy

Partitioning is **backend-specific**, not part of the core API.

### Why Not in Core API?

1. Different backends have vastly different partition capabilities
2. Some backends don't support partitioning at all (in-memory)
3. Keeps the abstract interface simple and focused
4. Backends can auto-optimize based on query filters

### Common Approach

Most persistent backends will partition by `data_version` fields because:
- **Immutable**: Data versions never change
- **Natural boundary**: Each version is independent
- **Query pattern**: Often filtering by specific versions
- **Incremental writes**: New versions append naturally

Backend configuration example (future):
```python
# Backend configures partitioning, not core API
delta_store = DeltaMetadataStore(
    table_uri="s3://bucket/metadata/my_feature",
    partition_cols=["data_version"],  # Backend-specific config
)

# Reads automatically benefit from partition pruning
df = delta_store.read_metadata(
    MyFeature,
    filters=pl.col("data_version").struct.field("default") == "abc123"
)
# Backend automatically prunes to relevant partitions
```

## Best Practices

### 1. Design for Immutability

Structure your pipelines to produce immutable outputs:
```python
# ✅ Good: Generate new data_version for each run
new_data = process_samples(samples, code_version=2)
store.write_metadata(MyFeature, new_data)

# ❌ Bad: Don't try to "update" existing data
# (This isn't supported - data is immutable)
```

### 2. Use Feature Classes

Pass `Feature` classes instead of `FeatureKey` strings:
```python
# ✅ Good: Type-safe, auto-resolves dependencies
store.read_metadata(VideoFeature)

# ⚠️ Works but less convenient: Manual key construction
store.read_metadata(FeatureKey(["video", "processing"]))
```

### 3. Leverage Fallback Chains

Structure your deployment to reuse stable upstream data:
```python
# ✅ Good: Dev reads from prod, writes locally
dev_store = SomeStore(fallback_stores=[prod_store])

# ❌ Bad: Recomputing everything in dev
dev_store = SomeStore()  # No fallback, must compute all deps
```

### 4. Check Existence Before Computing

Use `has_feature()` to avoid unnecessary computation:
```python
# ✅ Good: Check before computing
if not store.has_feature(MyFeature):
    # Feature doesn't exist, compute it
    df = compute_expensive_feature()
    store.write_metadata(MyFeature, df)
else:
    # Feature exists, just read it
    df = store.read_metadata(MyFeature)

# ❌ Less efficient: Relying on exception handling
try:
    df = store.read_metadata(MyFeature)
except FeatureNotFoundError:
    df = compute_expensive_feature()
    store.write_metadata(MyFeature, df)
```

### 5. Filter Early

Use Polars expressions to filter at the storage layer:
```python
# ✅ Good: Filter pushed to storage
df = store.read_metadata(
    MyFeature,
    filters=pl.col("sample_id").is_in([1, 2, 3])
)

# ⚠️ Less efficient: Filtering after reading all data
df = store.read_metadata(MyFeature)
df = df.filter(pl.col("sample_id").is_in([1, 2, 3]))
```

### 6. Schema Validation

Ensure your DataFrames have the required `data_version` column:
```python
# The store will validate on write
try:
    store.write_metadata(MyFeature, df_without_data_version)
except MetadataSchemaError:
    # Add data_version column first
    df = calculate_data_versions(df_without_data_version)
    store.write_metadata(MyFeature, df)
```

## Error Handling

Common exceptions:

```python
from metaxy.metadata_store import (
    FeatureNotFoundError,
    MetadataSchemaError,
    DependencyError,
)

# Feature doesn't exist
try:
    df = store.read_metadata(MyFeature, allow_fallback=False)
except FeatureNotFoundError:
    # Compute and materialize
    df = compute_feature()
    store.write_metadata(MyFeature, df)

# Invalid schema
try:
    store.write_metadata(MyFeature, invalid_df)
except MetadataSchemaError as e:
    print(f"Schema validation failed: {e}")

# Missing upstream dependencies
try:
    result = store.calculate_and_write_data_versions(
        MyFeature,
        sample_df,
        allow_upstream_fallback=False
    )
except DependencyError as e:
    print(f"Missing upstream feature: {e}")
```

## What's Next?

- **API Reference**: Detailed method signatures and parameters
- **Roadmap** (`docs/roadmap.md`): Future features like MaterializationCatalog, additional backends
- **Examples**: Complete end-to-end pipeline examples
- **Advanced Topics**: Distributed computing, optimization strategies
