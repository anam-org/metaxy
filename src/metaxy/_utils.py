import narwhals as nw
import polars as pl


def collect_to_polars(lazy_frame: nw.LazyFrame) -> pl.DataFrame:
    """Helper to collect a Narwhals LazyFrame and convert to Polars DataFrame.

    This handles all backend conversions (Polars, DuckDB/PyArrow, etc.) transparently.
    Used throughout tests for materializing query results.
    """
    return lazy_frame.collect().to_polars()
