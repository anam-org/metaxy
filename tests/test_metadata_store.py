"""Tests for metadata store."""

from collections.abc import Iterator

import polars as pl
import pytest

from metaxy import (
    ContainerDep,
    ContainerKey,
    ContainerSpec,
    Feature,
    FeatureDep,
    FeatureKey,
    FeatureRegistry,
    FeatureSpec,
)
from metaxy.metadata_store import (
    DependencyError,
    FeatureNotFoundError,
    InMemoryMetadataStore,
    MetadataSchemaError,
    StoreNotOpenError,
)


@pytest.fixture
def registry() -> Iterator[FeatureRegistry]:
    """Create a clean FeatureRegistry for testing."""
    reg = FeatureRegistry()
    with reg.use():
        yield reg


# Test fixtures - Define features for testing - they will use the registry from context


class UpstreamFeatureA(
    Feature,
    spec=FeatureSpec(
        key=FeatureKey(["upstream", "a"]),
        deps=None,
        containers=[
            ContainerSpec(key=ContainerKey(["frames"]), code_version=1),
            ContainerSpec(key=ContainerKey(["audio"]), code_version=1),
        ],
    ),
):
    pass


class UpstreamFeatureB(
    Feature,
    spec=FeatureSpec(
        key=FeatureKey(["upstream", "b"]),
        deps=None,
        containers=[
            ContainerSpec(key=ContainerKey(["default"]), code_version=1),
        ],
    ),
):
    pass


class DownstreamFeature(
    Feature,
    spec=FeatureSpec(
        key=FeatureKey(["downstream"]),
        deps=[
            FeatureDep(key=FeatureKey(["upstream", "a"])),
        ],
        containers=[
            ContainerSpec(
                key=ContainerKey(["default"]),
                code_version=1,
                deps=[
                    ContainerDep(
                        feature_key=FeatureKey(["upstream", "a"]),
                        containers=[ContainerKey(["frames"]), ContainerKey(["audio"])],
                    )
                ],
            ),
        ],
    ),
):
    pass


@pytest.fixture
def empty_store(registry: FeatureRegistry) -> Iterator[InMemoryMetadataStore]:
    """Empty in-memory store."""
    yield InMemoryMetadataStore()


@pytest.fixture
def populated_store(registry: FeatureRegistry) -> Iterator[InMemoryMetadataStore]:
    """Store with sample upstream data."""
    store = InMemoryMetadataStore()

    with store:
        # Add upstream feature A
        upstream_a_data = pl.DataFrame(
            {
                "sample_id": [1, 2, 3],
                "path": ["/data/1.mp4", "/data/2.mp4", "/data/3.mp4"],
                "data_version": [
                    {"frames": "hash_a1_frames", "audio": "hash_a1_audio"},
                    {"frames": "hash_a2_frames", "audio": "hash_a2_audio"},
                    {"frames": "hash_a3_frames", "audio": "hash_a3_audio"},
                ],
            }
        )
        store.write_metadata(UpstreamFeatureA, upstream_a_data)

    yield store


@pytest.fixture
def multi_env_stores(
    registry: FeatureRegistry,
) -> Iterator[dict[str, InMemoryMetadataStore]]:
    """Multi-environment store setup (prod, staging, dev)."""
    prod = InMemoryMetadataStore()
    staging = InMemoryMetadataStore(fallback_stores=[prod])
    dev = InMemoryMetadataStore(fallback_stores=[staging])

    with prod:
        # Populate prod with upstream data
        upstream_data = pl.DataFrame(
            {
                "sample_id": [1, 2, 3],
                "data_version": [
                    {"frames": "prod_hash1", "audio": "prod_hash1"},
                    {"frames": "prod_hash2", "audio": "prod_hash2"},
                    {"frames": "prod_hash3", "audio": "prod_hash3"},
                ],
            }
        )
        prod.write_metadata(UpstreamFeatureA, upstream_data)

    yield {"prod": prod, "staging": staging, "dev": dev}


# Basic CRUD Tests


def test_write_and_read_metadata(empty_store: InMemoryMetadataStore) -> None:
    """Test basic write and read operations."""
    with empty_store:
        metadata = pl.DataFrame(
            {
                "sample_id": [1, 2, 3],
                "data_version": [
                    {"frames": "hash1", "audio": "hash1"},
                    {"frames": "hash2", "audio": "hash2"},
                    {"frames": "hash3", "audio": "hash3"},
                ],
            }
        )

        empty_store.write_metadata(UpstreamFeatureA, metadata)
        result = empty_store.read_metadata(UpstreamFeatureA)

        assert len(result) == 3
        assert "sample_id" in result.columns
        assert "data_version" in result.columns


def test_write_invalid_schema(empty_store: InMemoryMetadataStore) -> None:
    """Test that writing without data_version column raises error."""
    with empty_store:
        invalid_df = pl.DataFrame(
            {
                "sample_id": [1, 2, 3],
                "path": ["/a", "/b", "/c"],
            }
        )

        with pytest.raises(MetadataSchemaError, match="data_version"):
            empty_store.write_metadata(UpstreamFeatureA, invalid_df)


def test_write_append(empty_store: InMemoryMetadataStore) -> None:
    """Test that writes are append-only."""
    with empty_store:
        df1 = pl.DataFrame(
            {
                "sample_id": [1, 2],
                "data_version": [
                    {"frames": "h1", "audio": "h1"},
                    {"frames": "h2", "audio": "h2"},
                ],
            }
        )

        df2 = pl.DataFrame(
            {
                "sample_id": [3, 4],
                "data_version": [
                    {"frames": "h3", "audio": "h3"},
                    {"frames": "h4", "audio": "h4"},
                ],
            }
        )

        empty_store.write_metadata(UpstreamFeatureA, df1)
        empty_store.write_metadata(UpstreamFeatureA, df2)

        result = empty_store.read_metadata(UpstreamFeatureA)
        assert len(result) == 4
        assert set(result["sample_id"].to_list()) == {1, 2, 3, 4}


def test_read_with_filters(populated_store: InMemoryMetadataStore) -> None:
    """Test reading with Polars filter expressions."""
    with populated_store:
        result = populated_store.read_metadata(
            UpstreamFeatureA, filters=pl.col("sample_id") > 1
        )

        assert len(result) == 2
        assert set(result["sample_id"].to_list()) == {2, 3}


def test_read_with_column_selection(populated_store: InMemoryMetadataStore) -> None:
    """Test reading specific columns."""
    with populated_store:
        result = populated_store.read_metadata(
            UpstreamFeatureA, columns=["sample_id", "data_version"]
        )

        assert set(result.columns) == {"sample_id", "data_version"}
        assert "path" not in result.columns


def test_read_nonexistent_feature(empty_store: InMemoryMetadataStore) -> None:
    """Test that reading nonexistent feature raises error."""
    with empty_store:
        with pytest.raises(FeatureNotFoundError):
            empty_store.read_metadata(UpstreamFeatureA)


# Feature Existence Tests


def test_has_feature_local(populated_store: InMemoryMetadataStore) -> None:
    """Test has_feature for local store."""
    with populated_store:
        assert populated_store.has_feature(UpstreamFeatureA, check_fallback=False)
        assert not populated_store.has_feature(UpstreamFeatureB, check_fallback=False)


def test_has_feature_with_fallback(
    multi_env_stores: dict[str, InMemoryMetadataStore],
) -> None:
    """Test has_feature checking fallback stores."""
    dev = multi_env_stores["dev"]
    staging = multi_env_stores["staging"]
    prod = multi_env_stores["prod"]

    with dev, staging, prod:
        # UpstreamFeatureA is in prod, not in dev
        assert not dev.has_feature(UpstreamFeatureA, check_fallback=False)
        assert dev.has_feature(UpstreamFeatureA, check_fallback=True)


def test_list_features(populated_store: InMemoryMetadataStore) -> None:
    """Test listing features."""
    with populated_store:
        features = populated_store.list_features()

        assert len(features) == 1
        assert any(f.to_string() == "upstream_a" for f in features)


def test_list_features_with_fallback(
    multi_env_stores: dict[str, InMemoryMetadataStore],
) -> None:
    """Test listing features including fallbacks."""
    dev = multi_env_stores["dev"]
    staging = multi_env_stores["staging"]
    prod = multi_env_stores["prod"]

    with dev, staging, prod:
        # Without fallback
        local_features = dev.list_features(include_fallback=False)
        assert len(local_features) == 0

        # With fallback
        all_features = dev.list_features(include_fallback=True)
        assert len(all_features) == 1
        assert any(f.to_string() == "upstream_a" for f in all_features)


# Fallback Store Tests


def test_read_from_fallback(multi_env_stores: dict[str, InMemoryMetadataStore]) -> None:
    """Test reading from fallback store."""
    dev = multi_env_stores["dev"]
    staging = multi_env_stores["staging"]
    prod = multi_env_stores["prod"]

    with dev, staging, prod:
        # Read from prod via fallback chain
        result = dev.read_metadata(UpstreamFeatureA, allow_fallback=True)
        assert len(result) == 3


def test_read_no_fallback(multi_env_stores: dict[str, InMemoryMetadataStore]) -> None:
    """Test that allow_fallback=False doesn't check fallback stores."""
    dev = multi_env_stores["dev"]

    with dev:
        with pytest.raises(FeatureNotFoundError):
            dev.read_metadata(UpstreamFeatureA, allow_fallback=False)


def test_write_to_dev_not_prod(
    multi_env_stores: dict[str, InMemoryMetadataStore],
) -> None:
    """Test that writes go to dev, not prod."""
    dev = multi_env_stores["dev"]
    prod = multi_env_stores["prod"]

    with dev, prod:
        new_data = pl.DataFrame(
            {
                "sample_id": [4, 5],
                "data_version": [{"default": "hash4"}, {"default": "hash5"}],
            }
        )

        dev.write_metadata(UpstreamFeatureB, new_data)

        # Should be in dev
        assert dev.has_feature(UpstreamFeatureB, check_fallback=False)

        # Should NOT be in prod
        assert not prod.has_feature(UpstreamFeatureB, check_fallback=False)


# Dependency Resolution Tests


def test_read_upstream_metadata(populated_store: InMemoryMetadataStore) -> None:
    """Test reading upstream dependencies."""
    with populated_store:
        upstream = populated_store.read_upstream_metadata(DownstreamFeature)

        assert "upstream_a" in upstream
        assert len(upstream["upstream_a"]) == 3
        assert "data_version" in upstream["upstream_a"].columns


def test_read_upstream_metadata_missing_dep(empty_store: InMemoryMetadataStore) -> None:
    """Test that missing dependencies raise error."""
    with empty_store:
        with pytest.raises(DependencyError, match="upstream_a"):
            empty_store.read_upstream_metadata(DownstreamFeature, allow_fallback=False)


def test_read_upstream_metadata_from_fallback(
    multi_env_stores: dict[str, InMemoryMetadataStore],
) -> None:
    """Test reading upstream from fallback stores."""
    dev = multi_env_stores["dev"]
    staging = multi_env_stores["staging"]
    prod = multi_env_stores["prod"]

    with dev, staging, prod:
        upstream = dev.read_upstream_metadata(DownstreamFeature, allow_fallback=True)

        assert "upstream_a" in upstream
        assert len(upstream["upstream_a"]) == 3


# Clear/Reset Tests


def test_clear_store(populated_store: InMemoryMetadataStore) -> None:
    """Test clearing store."""
    with populated_store:
        assert populated_store.has_feature(UpstreamFeatureA)

        populated_store.clear()

        assert not populated_store.has_feature(UpstreamFeatureA)
        assert len(populated_store.list_features()) == 0


def test_store_repr(empty_store: InMemoryMetadataStore) -> None:
    """Test string representation."""
    repr_str = repr(empty_store)

    assert "InMemoryMetadataStore" in repr_str
    assert "features=0" in repr_str


# Store Open/Close Tests


def test_store_not_open_write_raises(empty_store: InMemoryMetadataStore) -> None:
    """Test that writing to a closed store raises StoreNotOpenError."""
    metadata = pl.DataFrame(
        {
            "sample_id": [1, 2, 3],
            "data_version": [
                {"frames": "hash1", "audio": "hash1"},
                {"frames": "hash2", "audio": "hash2"},
                {"frames": "hash3", "audio": "hash3"},
            ],
        }
    )

    with pytest.raises(StoreNotOpenError, match="must be opened before use"):
        empty_store.write_metadata(UpstreamFeatureA, metadata)


def test_store_not_open_read_raises(populated_store: InMemoryMetadataStore) -> None:
    """Test that reading from a closed store raises StoreNotOpenError."""
    # Store is already closed after fixture setup
    with pytest.raises(StoreNotOpenError, match="must be opened before use"):
        populated_store.read_metadata(UpstreamFeatureA)


def test_store_not_open_has_feature_raises(
    populated_store: InMemoryMetadataStore,
) -> None:
    """Test that has_feature on a closed store raises StoreNotOpenError."""
    with pytest.raises(StoreNotOpenError, match="must be opened before use"):
        populated_store.has_feature(UpstreamFeatureA)


def test_store_not_open_list_features_raises(
    populated_store: InMemoryMetadataStore,
) -> None:
    """Test that list_features on a closed store raises StoreNotOpenError."""
    with pytest.raises(StoreNotOpenError, match="must be opened before use"):
        populated_store.list_features()


def test_store_context_manager_opens_and_closes(
    empty_store: InMemoryMetadataStore,
) -> None:
    """Test that context manager properly opens and closes store."""
    # Initially closed
    assert not empty_store._is_open

    with empty_store:
        # Should be open inside context
        assert empty_store._is_open

    # Should be closed after context
    assert not empty_store._is_open
