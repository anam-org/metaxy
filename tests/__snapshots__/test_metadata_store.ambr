# serializer version: 1
# name: test_branch_deployment_workflow
  list([
    '090b49a3fd8a3551ca7110fe225bc38bed44bce3012131cd0fda9448c2efc647',
    '4744f84bb3dca345cc65a65d36b16494c39f6149b54e7d12860ccfcfa5aa9b44',
    'b234af59fd88ea5c998f98ac43ce2c7372243ca4eaedaee78b199115af455817',
  ])
# ---
# name: test_calculate_and_write_data_versions
  list([
    'ab657b0686e72d308af0426ec5edff823d76645f367cb0f079ad494b3583b2e4',
    '28d35986093e005ef474f50dc31b71aa319097fff43ee77421d79033858cec56',
    'b6e3e8d060c63bcb7342ba961492d8f020cb9d86a5128f42246cbeaccc401429',
  ])
# ---
# name: test_calculate_deterministic
  'ab657b0686e72d308af0426ec5edff823d76645f367cb0f079ad494b3583b2e4'
# ---
# name: test_calculate_with_fallback_upstream
  list([
    '090b49a3fd8a3551ca7110fe225bc38bed44bce3012131cd0fda9448c2efc647',
    '4744f84bb3dca345cc65a65d36b16494c39f6149b54e7d12860ccfcfa5aa9b44',
  ])
# ---
